File: LeetCode/623. Add One Row to Tree/623.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root, val, depth):
        if depth == 1:
            new_root = TreeNode(val)
            new_root.left = root
            return new_root
        
        def addRow(node, current_depth):
            if node is None:
                return
            if current_depth == depth - 1:
                new_left = TreeNode(val, node.left, None)
                new_right = TreeNode(val, None, node.right)
                node.left = new_left
                node.right = new_right
            else:
                addRow(node.left, current_depth + 1)
                addRow(node.right, current_depth + 1)
        
        addRow(root, 1)
        return root

File: LeetCode/623. Add One Row to Tree/623.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if(depth==1){
            TreeNode newRoot=new TreeNode(val);
            newRoot.left=root;
            return newRoot;
        }
        addRow(root,1,val,depth);
        return root;
    }

        
    
    private void addRow(TreeNode node, int currentDepth, int val, int depth) {
        if(node==null){
            return;
        }
        if(currentDepth==depth-1){
            TreeNode newLeft=new TreeNode(val,node.left,null);
            TreeNode newRight=new TreeNode(val,null,node.right);
            node.left=newLeft;
            node.right=newRight;
        }else{
            addRow(node.left,currentDepth+1,val,depth);
            addRow(node.right,currentDepth+1,val,depth);
            
        }
    }
}